internal/controller/gitrepository_controller.go

```diff
@@ -1112,3 +1172,31 @@ func commitReference(obj *sourcev1.GitRepository, commit *git.Commit) string {
 	}
 	return commit.String()
 }
+
+// requiresVerification inspects a GitRepository's verification spec and its status
+// to determine whether the Git repository needs to be verified again. It does so by
+// first checking if the GitRepository has a verification spec. If it does, then
+// it returns true based on the following three conditions:
+//
+//   - If the object does not have a SourceVerifiedCondition in its status.
+//   - If the current reason for the SourceVerifiedCondition indicates that only the
+//     the tag has been verified earlier and the HEAD also needs to be verified now.
+//   - If the current reason for the SourceVerifiedCondition indicates that only the
+//     the HEAD has been verified earlier and the tag also needs to be verified now.
+func requiresVerification(obj *sourcev1.GitRepository) bool {
```

Comment:
On the fence about using conditions here versus a simple reflection of the latest observed value from the `.spec` in the status.

Reply:
are you talking about doing a `.status.observedGeneration == .metadata.generation`  and only proceed with (re-)verification if it returns false? if yes, we could do that but then we'd end up reverifying Git objects even when the change is unrelated, (a change in `.spec.interval` for example).

Reply:
No, more suggesting we should reflect the last observed configuration value in a new field, and use that to make the comparison. Instead of deciphering the conditions, which would make the (usually) arbitrary condition reason part of the official API.

Reply:
i'm not against having a new field, but i don't see how using reasons to figure out business logic makes it a part of the official API? we can change the reasons  to whatever we want later and then use it according to our needs. furthermore, i also believe that the new reasons proposed in the PR are more accurate than just "Succeeded".

Reply:
You are the consumer of your own object, and by making use of it in your business logic you are thus automatically making it part of the official API. As it now becomes something you consume to make decisions, and have to think about in the future to maintain backwards compatibility between controller versions.

I am not against the reasons being more informative, that's separate from using them in business logic.